url_matcher/matcher.py

```diff
@@ -0,0 +1,176 @@
+    include_matchers: List[PatternMatcher] = field(init=False)
+    exclude_matchers: List[PatternMatcher] = field(init=False)
+
+    def __post_init__(self):
+        self.include_matchers = [PatternMatcher(pattern) for pattern in self.patterns.include]
+        self.exclude_matchers = [PatternMatcher(pattern) for pattern in self.patterns.exclude]
+
+    def match(self, url: str) -> bool:
+        if self.include_matchers:
+            for include in self.include_matchers:
+                if include.match(url):
+                    break
+            else:
+                return False
+        for exclude in self.exclude_matchers:
+            if exclude.match(url):
+                return False
+        return True
+
+
+class IncludePatternsWithoutDomainError(ValueError):
+    def __init__(self, *args, id: Any, patterns: Patterns, wrong_patterns: List[str]):
+        super().__init__(*args)
+        self.id = id
+        self.patterns = patterns
+        self.wrong_patterns = wrong_patterns
+
+
+class URLMatcher:
+    def __init__(self, data: Union[Mapping[Any, Patterns], Iterable[Tuple[Any, Patterns]], None] = None):
+        """
+        A class that matches URLs against a list of patterns, returning
+        the identifier of the rule that matched the URL.
+
+        Example usage::
+
+            matcher = URLMatcher()
+            matcher.add_or_update(1, Patterns(include=["example.com/product"]))
+            matcher.add_or_update(2, Patterns(include=["other.com"]))
+
+            assert matcher.match("http://example.com/product/a_product.html") == 1
+            assert matcher.match("http://other.com/a_different_page") == 2
+
+        :param data: A map or a list of tuples with id, patterns pairs to
+                     initialize the object from
+        """
+        self.matchers_by_domain: Dict[str, List[PatternsMatcher]] = {}
+        self.patterns: Dict[Any, Patterns] = {}
+
+        if data:
+            items = data.items() if isinstance(data, Mapping) else data
+            for id, patterns in items:
+                self.add_or_update(id, patterns)
+
+    def add_or_update(self, id: Any, patterns: Patterns):
+        if not patterns.all_includes_have_domain() and not patterns.is_universal_pattern():
+            wrong_patterns = [p for p in patterns.get_includes_without_domain() if p]
+            raise IncludePatternsWithoutDomainError(
+                f"All include patterns must belong to a domain "
+                f"but the patterns {wrong_patterns} doesn't. "
+                f"For example, the include pattern '/product/* "
+                f"is invalid whereas the pattern 'example.com/product/*' isn't. "
+                f"The only exception is the empty pattern which matches everything "
+                f"and it is allowed. "
+                f"id: {id}.",
+                id=id,
+                patterns=patterns,
+                wrong_patterns=wrong_patterns,
+            )
+        if id in self.patterns:
```

Comment:
I would not be a problem in this context... but well, probably is better to avoid doing it. Done in https://github.com/zytedata/url-matcher/pull/2/commits/bd2c80bae459f9760fdac16f1ed05c44ad0a121f